core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableProject.java

```diff
@@ -128,36 +132,42 @@ public Result implement(EnumerableRelImplementor implementor, Prefer pref) {
     final Mappings.TargetMapping map =
         RelOptUtil.permutationIgnoreCast(
             getProjects(), getInput().getRowType());
-    if (checkIfSortIsDefinedOnNonTrivalExpr(map, collation)) {
-      return null;
+
+    List<RelFieldCollation> collationFieldsToDerive = new ArrayList<>();
+    for (RelFieldCollation rc : collation.getFieldCollations()) {
+      if (isCollationOnTrivialExpr(map, rc)) {
+        collationFieldsToDerive.add(rc);
+      }
```

Comment:
Let's see this query's plan:
```sql
select a, b*-1 as b, c from (select * from foo order by a, b, c limit 10) order by a, c;
```

Reply:
This is a good test case. Added (see `testSortProjectDerive5`)